hugolib/author.go

```diff
@@ -43,3 +71,189 @@ type Author struct {
 // - linkedin
 // - skype
 type AuthorSocial map[string]string
+
+// URL is a convenience function that provides the correct canonical URL
+// for a specific social network given a username. If an unsupported network
+// is requested, only the username is returned
+func (as AuthorSocial) URL(key string) string {
+	switch key {
+	case "github":
+		return fmt.Sprintf("https://github.com/%s", as[key])
+	case "facebook":
+		return fmt.Sprintf("https://www.facebook.com/%s", as[key])
+	case "twitter":
+		return fmt.Sprintf("https://twitter.com/%s", as[key])
+	case "googleplus":
+		isNumeric := onlyNumbersRegExp.Match([]byte(as[key]))
+		if isNumeric {
+			return fmt.Sprintf("https://plus.google.com/%s", as[key])
+		}
+		return fmt.Sprintf("https://plus.google.com/+%s", as[key])
+	case "pinterest":
+		return fmt.Sprintf("https://www.pinterest.com/%s/", as[key])
+	case "instagram":
+		return fmt.Sprintf("https://www.instagram.com/%s/", as[key])
+	case "youtube":
+		return fmt.Sprintf("https://www.youtube.com/user/%s", as[key])
+	case "linkedin":
+		return fmt.Sprintf("https://www.linkedin.com/in/%s", as[key])
+	default:
+		return as[key]
+	}
+}
+
+func mapToAuthorList(m map[string]interface{}) AuthorList {
+	authors := make(AuthorList, len(m))
+	for authorKey, data := range m {
+		authorMap, ok := data.(map[string]interface{})
+		if !ok {
+			continue
+		}
+		authors[authorKey] = mapToAuthor(authorMap)
+	}
+	return authors
+}
+
+func mapToAuthor(m map[string]interface{}) Author {
+	author := Author{}
+	for k, data := range m {
+		switch k {
+		case "givenName", "firstName":
+			author.GivenName = cast.ToString(data)
+			author.FirstName = author.GivenName
+		case "familyName", "lastName":
+			author.FamilyName = cast.ToString(data)
+			author.LastName = author.FamilyName
+		case "displayName":
+			author.DisplayName = cast.ToString(data)
+		case "thumbnail":
+			author.Thumbnail = cast.ToString(data)
+		case "image":
+			author.Image = cast.ToString(data)
+		case "shortBio":
+			author.ShortBio = cast.ToString(data)
+		case "bio":
+			author.Bio = cast.ToString(data)
+		case "email":
+			author.Email = cast.ToString(data)
+		case "social":
+			author.Social = normalizeSocial(cast.ToStringMapString(data))
+		case "params":
+			author.Params = cast.ToStringMapString(data)
+		}
+	}
+	return author
```

Comment:
We should assume that the users want's to use a custom display name instead of a hardcoded scheme. Furthermore, other cultures use different naming conventions. The Chinese [switch the given name and family name].